tests/DCPS/Compiler/xcdr/xcdr.cpp

```diff
@@ -0,0 +1,508 @@
+  {
+  }
+
+  DataView(const ACE_Message_Block& mb)
+  : data(mb.base())
+  , size(mb.length())
+  {
+  }
+
+  const char* const data;
+  const size_t size;
+};
+
+bool operator==(const DataView& a, const DataView& b)
+{
+  return a.size == b.size ? !std::memcmp(a.data, b.data, a.size) : false;
+}
+
+::testing::AssertionResult assert_DataView(
+  const char* a_expr, const char* b_expr,
+  const DataView& a, const DataView& b)
+{
+  if (a == b) {
+    return ::testing::AssertionSuccess();
+  }
+
+  std::string a_line;
+  std::stringstream astrm(to_hex_dds_string(a.data, a.size, '\n', 8));
+  std::string b_line;
+  std::stringstream bstrm(to_hex_dds_string(b.data, b.size, '\n', 8));
+  std::stringstream result;
+  bool got_a = true;
+  bool got_b = true;
+  bool diff = true;
+  while (got_a || got_b) {
+    result.width(16);
+    result << std::left;
+    if (got_a && std::getline(astrm, a_line, '\n')) {
+      result << a_line;
+    } else {
+      got_a = false;
+      result << ' ';
+    }
+    if (got_b && std::getline(bstrm, b_line, '\n')) {
+      result << " " << b_line;
+    } else {
+      got_b = false;
+    }
+    if (diff && a_line != b_line) {
+      result << " <- Different Starting Here";
+      diff = false;
+    }
+    result << std::endl;
+  }
+  return ::testing::AssertionFailure()
+    << a_expr << " (size " << a.size << ") isn't the same as "
+    << b_expr << " (size " << b.size << ").\n"
+    << a_expr << " is on the left, " << b_expr << " is on the right:\n"
+    << result.str();
+}
+
+template<typename TypeA, typename TypeB>
+void amalgam_serializer_test(
+  const Encoding& encoding, const DataView& expected_cdr, TypeA& value, TypeB& result)
+{
+  ACE_Message_Block buffer(1024);
+
+  // Serialize and Compare CDR
+  {
+    Serializer serializer(&buffer, encoding);
+    EXPECT_TRUE(serializer << value);
+    EXPECT_PRED_FORMAT2(assert_DataView, expected_cdr, buffer);
+  }
+
+  // Deserialize and Compare C++ Values
+  {
+    Serializer serializer(&buffer, encoding);
+    EXPECT_TRUE(serializer >> result);
+    EXPECT_PRED_FORMAT2(assert_values, value, result);
+  }
+}
+
+template<typename TypeA, typename TypeB>
+void amalgam_serializer_test(const Encoding& encoding, const DataView& expected_cdr)
+{
+  TypeA value;
+  set_values(value);
+  TypeB result;
+  amalgam_serializer_test<TypeA, TypeB>(encoding, expected_cdr, value, result);
+}
+
+template<typename Type>
+void serializer_test(const Encoding& encoding, const DataView& expected_cdr)
+{
+  amalgam_serializer_test<Type, Type>(encoding, expected_cdr);
+}
+
+template<typename Type>
+void baseline_checks(const Encoding& encoding, const DataView& expected_cdr)
+{
+  Type value;
+  EXPECT_EQ(serialized_size(encoding, value), expected_cdr.size);
+  // TODO(iguessthislldo): Does not work for XCDR1 and XCDR2 Mutable and XCDR2 Appendable
+  /* EXPECT_EQ(max_serialized_size(encoding, value), expected_cdr.size); */
+
+  serializer_test<Type>(encoding, expected_cdr);
```

Comment:
@mitza-oci Is there anything else that should be tested here other than `max_serialized_size`? Like `gen_skip_over`?

Reply:
This is for the future.

Reply:
I'm not sure about the "anything else" question.  I think that for `gen_skip_over` itself, that would be better tested as part of MetaStructTest using the "public" interface `MetaStruct<T>::getValue(Serializer&, const char*) const`.

Reply:
Yes you're right, `getValue` is what I actually want now that I'm actually looking at the code for it and it definitely needs to be updated for XCDR. I guess we could update the MetaStruct test, but it would need all of the non-union topic type test cases being defined here since `getValue` is going to have all of the same functionality going into `operator>>` for structs.